### ISSUE TITLE
Upgrade from Geany V1.24 to V1.26 breaks Build and Compile

### ISSUE THREAD

**Issue by Terry-euclid** [geany/geany#790]
Select build or compile from the Build menu is broken after upgrading to V1.26.

Downgrading back to 1.24 fixes this.

Terry

Microsoft Windows [Version 6.1.7601]
_________________________________________________________________________
V1.26 BUILD AND COMPILE BROKEN

15:58:11: Geany INFO        : Geany 1.26, English_United Kingdom.1252
15:58:11: Geany INFO        : GTK 2.24.10, GLib 2.28.8
15:58:11: Geany INFO        : System data dir: C:\Program Files\Geany\data
15:58:11: Geany INFO        : User config dir: C:\Users\terry.EUCLID\AppData\Roaming\geany
15:58:11: Geany INFO        : System plugin path: C:\Program Files\Geany\lib\geany
15:58:11: Geany INFO        : Added filetype Clojure (61).
15:58:11: Geany INFO        : Added filetype CUDA (62).
15:58:11: Geany INFO        : Added filetype Cython (63).
15:58:11: Geany INFO        : Added filetype Genie (64).
15:58:11: Geany INFO        : Added filetype Graphviz (65).
15:58:11: Geany INFO        : Added filetype JSON (66).
15:58:11: Geany INFO        : Added filetype Scala (67).
15:58:11: Geany INFO        : D:\projects\E2040\firmware\V1.0\com.c : C (UTF-8)
15:58:11: Geany INFO        : Loaded C:\Program Files\Geany\data\c99.tags (C), 1619 tag(s).
15:58:11: Geany INFO        : D:\projects\E2040\firmware\V1.0\calibrate.c : C (UTF-8)
15:58:14: Geany INFO        : build command spawning failed: The system cannot find the file specified
15:58:16: Geany INFO        : build command spawning failed: The directory name is invalid

_________________________________________________________________________
V1.24 BUILD AND COMPILE WORKING

15:59:30: Geany INFO        : Geany 1.24, English_United Kingdom.1252
15:59:30: Geany INFO        : GTK 2.16.6, GLib 2.20.5
15:59:30: Geany INFO        : System data dir: C:\Program Files\Geany\data
15:59:30: Geany INFO        : User config dir: C:\Users\terry.EUCLID\AppData\Roaming\geany
15:59:30: Geany INFO        : System plugin path: C:\Program Files\Geany\lib
15:59:30: Geany INFO        : Added filetype Clojure (58).
15:59:30: Geany INFO        : Added filetype CUDA (59).
15:59:30: Geany INFO        : Added filetype Cython (60).
15:59:30: Geany INFO        : Added filetype Genie (61).
15:59:30: Geany INFO        : Added filetype Go (62).
15:59:30: Geany INFO        : Added filetype Graphviz (63).
15:59:30: Geany INFO        : Added filetype Scala (64).
15:59:30: Geany INFO        : D:\projects\E2040\firmware\V1.0\com.c : C (UTF-8)
15:59:30: Geany INFO        : Loaded C:\Program Files\Geany\data\c99.tags (C), 1618 tag(s).
15:59:30: Geany INFO        : D:\projects\E2040\firmware\V1.0\calibrate.c : C (UTF-8)
15:59:35: Geany INFO        : WIN32: actually running command:
cmd.exe /S /C "run_splint "calibrate.c" >C:\Users\TERRY~1.EUC\AppData\Local\Temp\tmp_ODIE9X 2>C:\Users\TERRY~1.EUC\AppData\Local\Temp\tmp_PDIE9X"
15:59:37: Geany INFO        : WIN32: actually running command:
cmd.exe /S /C "build.bat -c "calibrate.c" >C:\Users\TERRY~1.EUC\AppData\Local\Temp\tmp_J59D9X 2>C:\Users\TERRY~1.EUC\AppData\Local\Temp\tmp_K59D9X"

**Comment by Terry-euclid** [comment]
OK done a bit more investigating and I think I have a fix.

My Build / Compile menu option was set as follows:

    command:                 build.bat -c "%f"   
    working directory        "%d"

Changing this to:

    command:                 cmd.exe /S /C build.bat -c "%f"   
    working directory        %d

Got things working.

I would be interested if someone could explain why this was changed as it looks to me like a step backwards.

What is the advantage of doing it the new way?

Terry

**Comment by b4n** [comment]
@Terry-euclid I believe the problem you have is only that your "working directory" option used quoting around `%d`, which it should not.  The command is a command, so might require some quoting, but the working directory is just a path.  If quoting was required on Windows before, it was unfortunately a bug.
So, please try using your old command, but simply removing the quotes around the working directory (e.g. just leaving `%d`).

> What is the advantage of doing it the new way?

Well… basically it's not a hack anymore :)  There was several issues regarding sub-process launching and communication on Windows, and they have been fixed (at least we believe so) in 1.25.
This should have been largely compatible with existing commands, but apparently we missed a corner case, sorry.  However, note that now build commands use quoting rules following Windows' conventions, which while mostly compatible has slight differences (i.e. in handling of `'`s).

Anyway, one glaring advantage of the "new way" is that Geany is now still responsive during build on Windows, and reports build command's output in real time, just like it does on other platforms.

**Comment by Terry-euclid** [comment]
@b4n Hi, thanks for the explanation.

There's no need to apologies about missing a corner case.   Geany is a great program and one of my favourites.  I love the fact I can use it under Windows and Linux.

I've tested it and found I do need the cmd.exe /S /C with V1.26 to make batch files run.   I've also found that when I go back to V1.24 I don't need the "" around the "working directory".

It was a long time ago when I set it up, but I believe I put the quotes around the "working directory" as a precaution, not because Geany didn't work, as I've had problems with other programs if the path has spaces in it.

Anyway, thanks for a great editor.

Terry
